README.md
# 秩序心海

提升「混沌心海」论坛用户体验的 Tampermonkey 脚本。

- [GitHub 源码仓库 - midniht/lawful-mcseas](https://github.com/midniht/lawful-mcseas) ![源码最新版本](https://img.shields.io/github/package-json/v/midniht/lawful-mcseas/main?logo=typescript&logoColor=fff&label=%E6%BA%90%E7%A0%81%E6%9C%80%E6%96%B0%E7%89%88%E6%9C%AC) ![最近更新](https://img.shields.io/github/last-commit/midniht/lawful-mcseas/gh-pages?logo=github&label=%E6%9C%80%E8%BF%91%E6%9B%B4%E6%96%B0)
- [Greasy Fork 发布地址 - 秩序心海](https://greasyfork.org/scripts/472436) ![脚本最新版本](https://img.shields.io/greasyfork/v/472436?logo=tampermonkey&label=%E8%84%9A%E6%9C%AC%E6%9C%80%E6%96%B0%E7%89%88%E6%9C%AC) ![脚本总安装量](https://img.shields.io/greasyfork/dt/472436?logo=tampermonkey&label=%E8%84%9A%E6%9C%AC%E6%80%BB%E5%AE%89%E8%A3%85%E9%87%8F)

## 功能

1. 自动格式化正文（文区的帖子一楼）内容：
   - 包括自定义字体和大小（原理是调用本地字体，只要电脑上安装了写上字体名就能用）；
   - 以及使用 [<ruby><rb>赫蹏</rb><rp>（</rp><rt>hè tí</rt><rp>）</rp></ruby>](https://github.com/sivan/heti) 自动进行排版优化，包括但不限于中英文混排美化以及更优雅的标点挤压等。\
      兼容 Discuz!X3.4 论坛自带的阅读模式。（沉浸式阅读模式有 bug：赫蹏 JS 未生效。由于是**新增节点**到 `div#append_parent` 这个容器里，需要在更新 DOM 后使用赫蹏重排。）
2. 自动跳过首页的「未验证邮箱」提示对话框。
3. 高亮显示当前地址（通过超链接）跳转到的楼层。
4. 可设置是否跳过「异地 IP 登录」安全提醒。
5. 可设置屏蔽「指定用户名」发布的「主题」。
6. 可设置「自动替换」正文中的关键词组（生效范围同「自动格式化」）。
7. 可设置在浏览帖子时隐藏所有用户的勋章。
8. 可设置是否在「自动格式化」时重新对正文进行分段并提行。

## Todo

- ~~优化高亮楼层的表现效果~~ 2023.8.6 done
- ~~沉浸式阅读模式下自动滚动 `document.documentElement.scrollTop`~~ 已放弃（长按鼠标中键：你在干嘛？）
- ~~将配置项与源码解耦，以免用户每次更新脚本后都需要重新设置~~ 2023.9.14 done
- ~~全文关键词替换（可以用来换主角名字，我觉得过于羞耻接受不能，但曾经听说 Pixiv 出过这功能，想必还是有市场的）~~ 2023.9.16 done
- ~~已读帖子变灰（理论上改下 `:visited` 伪类 CSS 颜色的事，但下意识感觉好像没有这么简单）~~ 2023.9.15 done 想多了
- 优化论坛自带的阅读模式（容器宽度、行间距等），做成真正的沉浸式
- ~~自定义字体 / 大小实时生效（现在是修改后自动刷新页面）~~ 2023.9.15 done
- 可选设置鼠标覆盖时 `:hover` 添加下划线（波浪线 `text-decoration-style: wavy;`）
- 高亮显示关注的用户
- 检测用户本地是否安装了某字体：
  1. 思路一（暴力）：直接随便找个字画出来（与 fallback 的默认字体）比较像素。
     - 优点：兼容性好。
     - 缺点：过于绿皮。
  2. 思路二（优雅）：使用 [`window.queryLocalFonts()`](https://developer.mozilla.org/en-US/docs/Web/API/Window/queryLocalFonts) 方法。
     - 优点：实现合理（精准而优雅）。
     - 缺点：兼容性过差（[Chrome 103 以上](https://caniuse.com/mdn-api_window_querylocalfonts) 才支持，目前只有 Chrome 和 Edge）。\
       不安全（貌似会弹警告），这是因为字体安装情况由于——不同 OS 版本、个人选择、软件自带「嫁妆」——带来的差异而具有高度的独特性，可以当作某种指纹使用（识别用户身份）。

## 开发

- [开发脚手架 lisonge/vite-plugin-monkey](https://github.com/lisonge/vite-plugin-monkey)
- [Tampermonkey 文档](https://www.tampermonkey.net/documentation.php)


CHANGELOG.md
# Changelog

> The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
> and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Alpha 前期测试]

### [**v0.5.2**] - _2023.9.17_

#### Fixed

- fix: 修复生成当前时间时月份错误
- fix: 修复「自动替换关键词」与上级设置没有聚合的问题

#### Changed

- refactor: 将工具类移动到单独的文件存放

### [**v0.5.1**] - _2023.9.16_

#### Added

- feat: 可设置是否在「自动格式化」时重新对正文进行分段并提行

#### Removed

- chore: 移除测试用设置菜单

### [**v0.5.0**] - _2023.9.16_

#### Added

- feat: 可设置在浏览帖子时隐藏所有用户的勋章

### [**v0.3.7**] - _2023.9.16_

#### Added

- feat: 可设置屏蔽「指定用户名」发布的「主题」
- feat: 可设置「自动替换」正文中的关键词组（生效范围同「自动格式化」）

#### Changed

- refactor: 梳理代码结构以获得更清晰的运行逻辑

### [**v0.3.6**] - _2023.9.15_

#### Added

- feat: 新增设置菜单快捷键

#### Fixed

- fix: 「自动格式化」功能现在对「资源区 > 小说」板块的帖子生效

### [**v0.3.5**] - _2023.9.15_

#### Changed

- improvement: 「自动格式化」功能修改后立刻生效（即时渲染）

### [**v0.3.4**] - _2023.9.15_

#### Added

- improvement: 优化样式 - 已读帖子变灰

### [**v0.3.3**] - _2023.9.14_

#### Deprecated

- refactor: 使用 TypeScript 完全重构整个脚本

#### Added

- feat: 可设置跳过异地 IP 登录提醒（由 [逝、死，存、生](https://mcseas.club/home.php?mod=space&uid=1019) 提出 [需求](https://mcseas.club/forum.php?mod=redirect&goto=findpost&ptid=50579&pid=1479122)）
- feat: 将配置项与代码解耦，新增脚本自带的设置菜单界面

#### Fixed

- fix: 「自动格式化」功能兼容 Discuz!X3.4 论坛自带的阅读模式
- fix: 「自动格式化」功能现在只对文区板块的帖子一楼生效（原创区包括图帖板块）

#### Changed

- improvement: 「自动关闭」等功能执行策略采用持续轮询（之前是等待加载时机）

### [**v0.1.5**] - _2023.8.6_

#### Changed

- improvement: 优化「高亮楼层的表现效果（简单改变背景颜色 → 带 `:hover` 动效的卡片样式）

### [**v0.1.4**] - _2023.8.6_

#### Added

- feat: 高亮显示当前地址（通过超链接）跳转到的楼层（由 [eliaaaa](https://mcseas.club/home.php?mod=space&uid=20101) 提出 [需求](https://mcseas.club/forum.php?mod=redirect&goto=findpost&ptid=50579&pid=1466392)）

### [**v0.1.3**] - _2023.8.5_

#### Added

- feat: 「自动格式化」功能可设置自定义正文（原创区帖子一楼）字体大小

#### Fixed

- fix: 兼容某些开头缺少「字数统计（论坛自带功能）」的旧版本帖子

### [**v0.1.2**] - _2023.8.5_

#### Added

- feat: 自动跳过首页的「未验证邮箱」提示对话框

### [**v0.1.1**] - _2023.8.4_

#### Added

- feat: 引入 [赫蹏](https://github.com/sivan/heti) 自动进行排版优化，包括但不限于中英文混排美化以及更优雅的标点挤压等


src/utils.ts
import { GM_log, GM_setValue, GM_getValue } from "$";
import pkg from "../package.json";

class Utils {
  current(): string {
    const t = new Date();
    const pad = (
      pad_target: any,
      pad_length: number = 2,
      pad_string: string = "0"
    ): string => pad_target.toString().padStart(pad_length, pad_string);
    return (
      [pad(t.getFullYear(), 4), pad(t.getMonth() + 1), pad(t.getDate())].join(
        "-"
      ) +
      " " +
      [pad(t.getHours()), pad(t.getMinutes()), pad(t.getSeconds())].join(":")
    );
  }

  debug(...args: any[]): void {
    GM_log(`${this.current()} | DEBUG |`, ...args);
  }

  log(...args: any[]): void {
    console.log(`${this.current()} [秩序心海]`, ...args);
  }

  // match_reg(text: string, regexp: RegExp): string | null {
  //   const match_result = [...text.matchAll(regexp)];
  //   return match_result.length > 0 ? match_result[0][1] : null;
  // }
}

class Config {
  readonly version: string = `${pkg.version}-alpha`;
  auto_format!: boolean;
  auto_format_with_segment!: boolean;
  font_name!: string;
  font_size!: number;
  only_format_lz!: boolean;
  block_ip_warning!: boolean;
  display_user_medal!: boolean;
  block_email_warning!: boolean;
  data!: {
    user_blacklist: string[];
    replace_pair_list: { [key: string]: string };
  };

  constructor() {
    this.load();
  }

  load(): void {
    this.auto_format = GM_getValue("cfg_auto_format") !== "false";
    this.auto_format_with_segment =
      GM_getValue("cfg_auto_format_with_segment") !== "false";
    this.font_name = GM_getValue("cfg_font_name") || "小赖字体 等宽 SC";
    // "LXGW Bright", "LXGW Wenkai Mono", "悠哉字体", "小赖字体 等宽 SC"
    this.font_size = Number(GM_getValue("cfg_font_size"));
    this.font_size = this.font_size > 0 ? this.font_size : 16;
    this.only_format_lz = GM_getValue("cfg_only_format_lz") !== "false";
    this.block_ip_warning = GM_getValue("cfg_block_ip_warning") === "true";
    this.display_user_medal = GM_getValue("cfg_display_user_medal") !== "false";
    this.block_email_warning =
      GM_getValue("cfg_block_email_warning") === "true";
    this.data = {
      user_blacklist: JSON.parse(GM_getValue("data_user_blacklist") ?? "[]"),
      replace_pair_list: JSON.parse(
        GM_getValue("data_replace_pair_list") ?? "{}"
      ),
    };
  }

  save(key: string, value: string): void {
    GM_setValue(`cfg_${key}`, value);
    this.load();
  }
}

type MenuIdMapKeys =
  | "auto_format"
  | "auto_format_with_segment"
  | "format_font_name"
  | "format_font_size"
  | "switch_ip_warning"
  | "display_user_medal"
  | "go_to_report"
  | "edit_replace_pair"
  | "reset_config";
type MenuIdMap = {
  [key in MenuIdMapKeys]: string;
};
let menu_id_map: MenuIdMap = {
  auto_format: "",
  auto_format_with_segment: "",
  format_font_name: "",
  format_font_size: "",
  switch_ip_warning: "",
  display_user_medal: "",
  go_to_report: "",
  edit_replace_pair: "",
  reset_config: "",
};

type StyleNodes = {
  custom_view_list: HTMLStyleElement | undefined;
  custom_format_font: HTMLStyleElement | undefined;
  custom_highlight_floor: HTMLStyleElement | undefined;
  heti: HTMLStyleElement | undefined;
};
let style_nodes: StyleNodes = {
  custom_view_list: undefined,
  custom_format_font: undefined,
  custom_highlight_floor: undefined,
  heti: undefined,
};

export { Utils, Config, menu_id_map, style_nodes };


vite.config.ts
import { defineConfig } from "vite";
import monkey, { util, cdn } from "vite-plugin-monkey";
// https://github.com/unplugin/unplugin-auto-import
// import AutoImport from "unplugin-auto-import/vite";

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [
    // AutoImport({
    //   imports: [util.unimportPreset], // 按需自动导入 GM_api
    // }),
    monkey({
      entry: "src/main.ts",
      userscript: {
        name: { en: "Lawful MC seas", "zh-CN": "秩序心海" },
        namespace: "https://mcseas.club/home.php?mod=space&uid=95082",
        version: "0.5.2-alpha",
        description: {
          en: "Improve the user experience of mcseas.",
          "zh-CN": "改善「混沌心海」论坛的使用体验。",
        },
        author: "miyoi",
        homepage: "https://mcseas.club/forum.php?mod=viewthread&tid=50579",
        homepageURL: "https://github.com/midniht/lawful-mcseas",
        license: "MIT",
        match: ["*://mcseas.club/*"],
        icon: "https://mcseas.club/favicon.ico",
        require: ["https://unpkg.com/heti/umd/heti-addon.min.js"],
        resource: { css: "https://unpkg.com/heti/umd/heti.min.css" },
        "run-at": "document-end",
      },
      // build: {
      //   externalGlobals: {
      //     heti: cdn.unpkg("heti", "umd/heti-addon.min.js"),
      //   },
      //   externalResource: {
      //     "heti/umd/heti.min.css": cdn.unpkg(),
      //   },
      // },
      // server: { mountGmApi: true }, // 全局变量导入 GM_api
    }),
  ],
  build: {
    // minify: true,
  },
});


lawful-mcseas.user.js
// ==UserScript==
// @name               lawful-mcseas
// @name:en            Lawful MC seas
// @name:zh-CN         秩序心海
// @namespace          https://mcseas.club/home.php?mod=space&uid=95082
// @version            0.5.2-alpha
// @author             miyoi
// @description:en     Improve the user experience of mcseas.
// @description:zh-CN  改善「混沌心海」论坛的使用体验。
// @license            MIT
// @icon               https://mcseas.club/favicon.ico
// @homepage           https://mcseas.club/forum.php?mod=viewthread&tid=50579
// @homepageURL        https://github.com/midniht/lawful-mcseas
// @match              *://mcseas.club/*
// @require            https://unpkg.com/heti/umd/heti-addon.min.js
// @resource           css  https://unpkg.com/heti/umd/heti.min.css
// @grant              GM_addStyle
// @grant              GM_deleteValue
// @grant              GM_getResourceText
// @grant              GM_getValue
// @grant              GM_listValues
// @grant              GM_log
// @grant              GM_openInTab
// @grant              GM_registerMenuCommand
// @grant              GM_setValue
// @grant              GM_unregisterMenuCommand
// @run-at             document-end
// ==/UserScript==

(function () {
  'use strict';

  var __defProp = Object.defineProperty;
  var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
  var __publicField = (obj, key, value) => {
    __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
    return value;
  };
  var _a, _b, _c, _d, _e, _f, _g;
  var _GM_addStyle = /* @__PURE__ */ (() => typeof GM_addStyle != "undefined" ? GM_addStyle : void 0)();
  var _GM_deleteValue = /* @__PURE__ */ (() => typeof GM_deleteValue != "undefined" ? GM_deleteValue : void 0)();
  var _GM_getResourceText = /* @__PURE__ */ (() => typeof GM_getResourceText != "undefined" ? GM_getResourceText : void 0)();
  var _GM_getValue = /* @__PURE__ */ (() => typeof GM_getValue != "undefined" ? GM_getValue : void 0)();
  var _GM_listValues = /* @__PURE__ */ (() => typeof GM_listValues != "undefined" ? GM_listValues : void 0)();
  var _GM_log = /* @__PURE__ */ (() => typeof GM_log != "undefined" ? GM_log : void 0)();
  var _GM_openInTab = /* @__PURE__ */ (() => typeof GM_openInTab != "undefined" ? GM_openInTab : void 0)();
  var _GM_registerMenuCommand = /* @__PURE__ */ (() => typeof GM_registerMenuCommand != "undefined" ? GM_registerMenuCommand : void 0)();
  var _GM_setValue = /* @__PURE__ */ (() => typeof GM_setValue != "undefined" ? GM_setValue : void 0)();
  var _GM_unregisterMenuCommand = /* @__PURE__ */ (() => typeof GM_unregisterMenuCommand != "undefined" ? GM_unregisterMenuCommand : void 0)();
  var _monkeyWindow = /* @__PURE__ */ (() => window)();
  const name = "lawful-mcseas";
  const version = "0.5.2";
  const type = "module";
  const scripts = {
    dev: "vite",
    build: "tsc && vite build",
    preview: "vite preview"
  };
  const devDependencies = {
    typescript: "^5.1.6",
    "unplugin-auto-import": "^0.16.6",
    vite: "^4.4.8",
    "vite-plugin-monkey": "^3.4.0"
  };
  const pkg = {
    name,
    "private": true,
    version,
    type,
    scripts,
    devDependencies
  };
  class Utils {
    current() {
      const t = /* @__PURE__ */ new Date();
      const pad = (pad_target, pad_length = 2, pad_string = "0") => pad_target.toString().padStart(pad_length, pad_string);
      return [pad(t.getFullYear(), 4), pad(t.getMonth() + 1), pad(t.getDate())].join(
        "-"
      ) + " " + [pad(t.getHours()), pad(t.getMinutes()), pad(t.getSeconds())].join(":");
    }
    debug(...args) {
      _GM_log(`${this.current()} | DEBUG |`, ...args);
    }
    log(...args) {
      console.log(`${this.current()} [秩序心海]`, ...args);
    }
    // match_reg(text: string, regexp: RegExp): string | null {
    //   const match_result = [...text.matchAll(regexp)];
    //   return match_result.length > 0 ? match_result[0][1] : null;
    // }
  }
  class Config {
    constructor() {
      __publicField(this, "version", `${pkg.version}-alpha`);
      __publicField(this, "auto_format");
      __publicField(this, "auto_format_with_segment");
      __publicField(this, "font_name");
      __publicField(this, "font_size");
      __publicField(this, "only_format_lz");
      __publicField(this, "block_ip_warning");
      __publicField(this, "display_user_medal");
      __publicField(this, "block_email_warning");
      __publicField(this, "data");
      this.load();
    }
    load() {
      this.auto_format = _GM_getValue("cfg_auto_format") !== "false";
      this.auto_format_with_segment = _GM_getValue("cfg_auto_format_with_segment") !== "false";
      this.font_name = _GM_getValue("cfg_font_name") || "小赖字体 等宽 SC";
      this.font_size = Number(_GM_getValue("cfg_font_size"));
      this.font_size = this.font_size > 0 ? this.font_size : 16;
      this.only_format_lz = _GM_getValue("cfg_only_format_lz") !== "false";
      this.block_ip_warning = _GM_getValue("cfg_block_ip_warning") === "true";
      this.display_user_medal = _GM_getValue("cfg_display_user_medal") !== "false";
      this.block_email_warning = _GM_getValue("cfg_block_email_warning") === "true";
      this.data = {
        user_blacklist: JSON.parse(_GM_getValue("data_user_blacklist") ?? "[]"),
        replace_pair_list: JSON.parse(
          _GM_getValue("data_replace_pair_list") ?? "{}"
        )
      };
    }
    save(key, value) {
      _GM_setValue(`cfg_${key}`, value);
      this.load();
    }
  }
  let menu_id_map = {
    auto_format: "",
    auto_format_with_segment: "",
    format_font_name: "",
    format_font_size: "",
    switch_ip_warning: "",
    display_user_medal: "",
    go_to_report: "",
    edit_replace_pair: "",
    reset_config: ""
  };
  let style_nodes = {
    custom_view_list: void 0,
    custom_format_font: void 0,
    custom_highlight_floor: void 0,
    heti: void 0
  };
  const utils = new Utils();
  const setting = new Config();
  const rerender_auto_format = (include_heti = true) => {
    var _a2, _b2, _c2, _d2;
    if (setting.auto_format) {
      if (include_heti) {
        (_a2 = style_nodes.heti) == null ? void 0 : _a2.remove();
        style_nodes.heti = _GM_addStyle(_GM_getResourceText("css"));
        utils.debug(
          `自动格式化正文 赫蹏 更新 CSS id: style#${style_nodes.heti.id}`
        );
        window.onload = () => {
          const heti = new _monkeyWindow.Heti(".heti, .heti-parent .pcb");
          heti.autoSpacing();
        };
      }
      (_b2 = style_nodes.custom_format_font) == null ? void 0 : _b2.remove();
      style_nodes.custom_format_font = _GM_addStyle(`
      .heti, .heti-parent .pcb { font-family: "${setting.font_name}", "Helvetica Neue", helvetica, arial, "Heti Hei", sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol"; font-size: ${setting.font_size}px; }
      .t_f > div:not(.blockcode) { margin-bottom: .75em; }
    `);
      utils.debug(
        `自动格式化正文 自定义字体 更新 CSS id: style#${style_nodes.custom_format_font.id}`
      );
    } else {
      (_c2 = style_nodes.heti) == null ? void 0 : _c2.remove();
      (_d2 = style_nodes.custom_format_font) == null ? void 0 : _d2.remove();
    }
  };
  const recreate_menu_command = () => {
    for (let [menu_key, menu_id] of Object.entries(menu_id_map)) {
      if (menu_id) {
        _GM_unregisterMenuCommand(menu_id);
        menu_id_map[menu_key] = "";
      }
    }
    menu_id_map.auto_format = _GM_registerMenuCommand(
      (setting.auto_format ? "✔️ 已启用" : "❌ 已禁用") + "自动格式化正文",
      () => {
        setting.save("auto_format", setting.auto_format ? "false" : "true");
        utils.log(
          (setting.auto_format ? "✔️ 已启用" : "❌ 已禁用") + "自动格式化正文"
        );
        rerender_auto_format();
        recreate_menu_command();
      },
      "r"
    );
    if (setting.auto_format) {
      menu_id_map.auto_format_with_segment = _GM_registerMenuCommand(
        "　├─ " + (setting.auto_format_with_segment ? "✔️ 已启用" : "❌ 已禁用") + "自动重新分段",
        () => {
          setting.save(
            "auto_format_with_segment",
            setting.auto_format_with_segment ? "false" : "true"
          );
          utils.log(
            (setting.auto_format_with_segment ? "✔️ 已启用" : "❌ 已禁用") + "自动重新分段"
          );
          window.location.reload();
        },
        "r"
      );
      menu_id_map.format_font_name = _GM_registerMenuCommand(
        `　├─ 🗚 字体: "${setting.font_name}"`,
        () => {
          const font_name = prompt(
            "请输入自定义字体名称（需要电脑里已经安装好的字体）：",
            setting.font_name
          );
          if (font_name) {
            setting.save("font_name", font_name);
            utils.log(`字体设置为: "${font_name}"`);
            rerender_auto_format(false);
            recreate_menu_command();
          } else {
            utils.debug("用户取消输入");
          }
        },
        "f"
      );
      menu_id_map.format_font_size = _GM_registerMenuCommand(
        `　├─ 🗚 字体大小: ${setting.font_size} px`,
        () => {
          const font_size = Number(
            prompt(
              "请输入自定义字体大小，单位为像素(px)：",
              String(setting.font_size)
            )
          );
          if (font_size > 0) {
            setting.save("font_size", String(font_size));
            utils.log(`字体大小设置为：${font_size} px`);
            rerender_auto_format(false);
            recreate_menu_command();
          } else {
            utils.debug("用户取消输入");
          }
        },
        "s"
      );
      menu_id_map.edit_replace_pair = _GM_registerMenuCommand(
        "　└─ 🎭 设置自动替换关键词",
        () => {
          const current_replace_pair_str = Object.entries(
            setting.data.replace_pair_list
          ).map(([key, value]) => `${key}-${value}`).join(", ");
          const replace_pair_str = prompt(
            "请输入自动替换的关键词组：\n（格式为 `被替换词-替换词`，多个词组用英文逗号 `,` 分开）\n\n注意：只推荐替换中文全角字符，如果替换常见英文字符极有可能会导致乱码。若出现乱码请重新在此处设置以调试效果。",
            current_replace_pair_str || "“-「, ”-」, ‘-『, ’-』"
          );
          if (replace_pair_str !== null) {
            setting.data.replace_pair_list = {};
            _GM_setValue(`data_replace_pair_list`, "{}");
            if (replace_pair_str === "") {
              utils.debug("自动替换词组已清空");
              alert("自动替换词组已清空。");
            } else {
              for (const pair_str of replace_pair_str.split(",").map((pair) => pair.trim())) {
                const [pattern, replacement] = pair_str.split("-").map((keyword) => keyword.trim());
                setting.data.replace_pair_list[pattern] = replacement;
              }
              _GM_setValue(
                `data_replace_pair_list`,
                JSON.stringify(setting.data.replace_pair_list)
              );
              utils.debug("新的自动替换词组:", setting.data.replace_pair_list);
              alert("自动替换词组更新完成。");
            }
            window.location.reload();
          } else {
            utils.debug("用户取消输入");
          }
        }
      );
    }
    menu_id_map.switch_ip_warning = _GM_registerMenuCommand(
      (setting.block_ip_warning ? "❌ 已禁用" : "✔️ 已启用") + "异地 IP 登录提醒",
      () => {
        setting.save(
          "block_ip_warning",
          setting.block_ip_warning ? "false" : "true"
        );
        utils.log(
          (setting.block_ip_warning ? "❌ 已禁用" : "✔️ 已启用") + "异地 IP 登录提醒"
        );
        recreate_menu_command();
      }
    );
    menu_id_map.display_user_medal = _GM_registerMenuCommand(
      (setting.display_user_medal ? "✔️ 显示" : "❌ 隐藏") + "所有用户勋章",
      () => {
        setting.save(
          "display_user_medal",
          setting.display_user_medal ? "false" : "true"
        );
        utils.log(
          (setting.display_user_medal ? "✔️ 已显示" : "❌ 已隐藏") + "所有用户勋章"
        );
        window.location.reload();
      }
    );
    menu_id_map.go_to_report = _GM_registerMenuCommand(
      "🆕 提出需求 & 反馈 BUG",
      () => {
        _GM_openInTab(
          "https://mcseas.club/forum.php?mod=viewthread&tid=50579",
          false
        );
      }
    );
    menu_id_map.reset_config = _GM_registerMenuCommand("🆑 重置所有配置项", () => {
      const confirm_result = confirm(`确认清空保存的所有配置项？`);
      if (confirm_result == true) {
        for (const key of _GM_listValues()) {
          _GM_deleteValue(key);
        }
        alert("所有配置项均已清空。");
        window.location.reload();
      }
    });
  };
  window.block_user = (user_name) => {
    const confirm_result = confirm(
      `确认屏蔽 ${user_name} 发布的主题？
（屏蔽数据仅保存在本地，可以通过重置插件设置来取消屏蔽）`
    );
    if (confirm_result == true) {
      setting.data.user_blacklist.push(user_name);
      _GM_setValue(
        `data_user_blacklist`,
        JSON.stringify([...new Set(setting.data.user_blacklist)])
      );
      alert(`从今开始屏蔽用户 ${user_name} 发布的主题。`);
    }
  };
  utils.log("脚本当前版本:", setting.version);
  utils.log("启用自动格式化正文:", setting.auto_format);
  utils.log(`自动格式化时 正文字体: "${setting.font_name}"`);
  utils.log(`自动格式化时 正文字体大小: ${setting.font_size} px`);
  utils.log("自动格式化时 替换关键词:", setting.data.replace_pair_list);
  utils.log("自动格式化时 只格式化一楼:", setting.only_format_lz);
  utils.log("自动屏蔽异地 IP 登录提醒:", setting.block_ip_warning);
  utils.log("屏蔽用户黑名单:", setting.data.user_blacklist);
  recreate_menu_command();
  if (setting.block_ip_warning) {
    let block_ip_warning = function() {
      counter += 1;
      if (counter > 50)
        clearInterval(block_ip_timer);
      let close_button_nodes = document.querySelectorAll("#ip_notice .bm_h a");
      if (close_button_nodes.length === 1) {
        close_button_nodes[0].click();
        clearInterval(block_ip_timer);
      }
    };
    let block_ip_timer = setInterval(() => {
      setTimeout(block_ip_warning, 0);
    }, 50);
    let counter = 0;
  }
  let view_mode = new URL(window.location.href).searchParams.get("mod") ?? "";
  switch (view_mode) {
    case "forumdisplay":
      let thread_nodes = document.querySelectorAll(
        "#threadlisttableid tbody"
      );
      thread_nodes.forEach((thread_node) => {
        if (thread_node.id.startsWith("normalthread_")) {
          const thread_title_node = thread_node.querySelector(".cony_listzt_top a.cony_listzt_topbt");
          const thread_id = Number(
            new URL((thread_title_node == null ? void 0 : thread_title_node.href) ?? "").searchParams.get("tid")
          );
          const thread_title = (thread_title_node == null ? void 0 : thread_title_node.innerText) ?? "";
          const thread_author_node = thread_node.querySelector(".cony_listzt_bottom .z a");
          const uid = Number(
            new URL((thread_title_node == null ? void 0 : thread_title_node.href) ?? "").searchParams.get("uid")
          );
          const author_name = (thread_author_node == null ? void 0 : thread_author_node.innerText) ?? "";
          if (author_name && setting.data.user_blacklist.includes(author_name)) {
            thread_node.remove();
            utils.debug(
              `已屏蔽黑名单用户 "${author_name}"(uid:${uid}) 的主题 "${thread_title}"(tid:${thread_id})`
            );
          }
        }
      });
      style_nodes.custom_view_list = _GM_addStyle(
        `.tl th a:visited, .tl td.fn a:visited { color: #ccc; }`
      );
      utils.debug(
        `添加自定义样式 新增 CSS id: style#${(_a = style_nodes.custom_view_list) == null ? void 0 : _a.id}`
      );
      break;
    case "viewthread":
      if (!setting.display_user_medal) {
        let medal_nodes = document.querySelectorAll(".md_ctrl");
        medal_nodes.forEach((node) => {
          node.remove();
        });
      }
      let follow_node = document.querySelector("#follow_li");
      if (follow_node) {
        follow_node.style.width = "65px";
        if (follow_node.parentNode) {
          let list_item_node = document.createElement("li");
          list_item_node.setAttribute("class", "pm2");
          list_item_node.setAttribute(
            "style",
            "background-image: url('static/image/common/locked.gif');"
          );
          let block_user_button_node = document.createElement("a");
          block_user_button_node.setAttribute("href", "javascript:;");
          block_user_button_node.setAttribute("title", "屏蔽TA");
          block_user_button_node.setAttribute("class", "xi2");
          block_user_button_node.setAttribute(
            "onclick",
            `block_user("${(_d = (_c = (_b = follow_node.parentNode) == null ? void 0 : _b.parentNode) == null ? void 0 : _c.querySelector(
            ".authi a"
          )) == null ? void 0 : _d.innerText}")`
          );
          block_user_button_node.innerHTML = "屏蔽TA";
          list_item_node.append(block_user_button_node);
          follow_node.parentNode.append(list_item_node);
        }
      }
      const highlight_floor_pid = new URL(window.location.href).hash;
      if (highlight_floor_pid.startsWith("#pid")) {
        let highlight_floor_node = document.querySelector(highlight_floor_pid);
        if (highlight_floor_node) {
          (_e = highlight_floor_node.querySelector("td.plc")) == null ? void 0 : _e.classList.add("highlight-card");
          style_nodes.custom_highlight_floor = _GM_addStyle(
            `.highlight-card { background-color: #dedbcc; color: #363636; -moz-box-shadow: 0.075rem 0.125rem 0.25rem rgba(0, 0, 0, 0.5); -webkit-box-shadow: 0.075rem 0.125rem 0.25rem rgba(0, 0, 0, 0.5); box-shadow: 0.075rem 0.125rem 0.25rem rgba(0, 0, 0, 0.5); } .highlight-card:hover { -webkit-box-shadow: 0 0.325rem 1.75rem rgba(0, 0, 0, 0.3); -moz-box-shadow: 0 0.325rem 1.75rem rgba(0, 0, 0, 0.3); box-shadow: 0 0.325rem 1.75rem rgba(0, 0, 0, 0.3); }`
          );
          utils.debug(
            `高亮当前楼层 新增 CSS id: style#${style_nodes.custom_highlight_floor.id}`
          );
        }
      }
      (_f = document.querySelector("#append_parent")) == null ? void 0 : _f.classList.add("heti-parent");
      let breadcrumb_nodes = document.querySelectorAll("div#pt > div.z > a");
      if (breadcrumb_nodes.length > 4 && ["原创文学", "审核区", "小说"].includes(breadcrumb_nodes[3].innerText)) {
        let post_nodes = document.querySelectorAll("table.plhin td.plc");
        for (let i = 0, len = post_nodes.length; i < len; i++) {
          if (setting.only_format_lz && ((_g = post_nodes[i].querySelector("div.pi > strong > a")) == null ? void 0 : _g.innerText.trim()) !== "战列舰")
            break;
          let post_node = post_nodes[i].querySelector(".t_f");
          if (post_node) {
            post_node.classList.add("heti");
            if (setting.auto_format_with_segment) {
              post_node.innerHTML = post_node.innerHTML.split("\n").map((paragraph) => {
                const formatted_paragraph = paragraph.replaceAll("&nbsp;", " ").trim();
                return formatted_paragraph ? `　　${formatted_paragraph}` : "";
              }).join("\n");
              post_node.querySelectorAll("div, p, span").forEach((node) => {
                if (node.style.display !== "none") {
                  node.innerHTML = "　　" + node.innerHTML.replaceAll("&nbsp;", " ").trim();
                }
              });
            }
            for (let [pattern, replacement] of Object.entries(
              setting.data.replace_pair_list
            )) {
              post_node.innerHTML = post_node.innerHTML.replaceAll(
                pattern,
                replacement
              );
            }
          }
        }
      }
      rerender_auto_format();
      break;
    default:
      let block_email_warning = function() {
        counter += 1;
        if (counter > 50)
          clearInterval(block_email_timer);
        let skip_button_node = document.querySelector(
          "button#fwin_dialog_submit"
        );
        if (skip_button_node) {
          skip_button_node.click();
          clearInterval(block_email_timer);
        }
      };
      let block_email_timer = setInterval(() => {
        setTimeout(block_email_warning, 0);
      }, 50);
      let counter = 0;
  }

})();